apiserver/common/networkingcommon/linklayer.go

```diff
@@ -0,0 +1,157 @@
+package networkingcommon
+
+import (
+	"github.com/juju/collections/set"
+	"github.com/juju/errors"
+	"gopkg.in/mgo.v2/txn"
+
+	"github.com/juju/juju/core/network"
+)
+
+// LinkLayerDevice describes a single layer-2 network device.
+type LinkLayerDevice interface {
+	// MACAddress is the hardware address of the device.
+	MACAddress() string
+
+	// Name is the name of the device.
+	Name() string
+
+	// ProviderID returns the provider-specific identifier for this device.
+	ProviderID() network.Id
+
+	// SetProviderIDOps returns the operations required to set the input
+	// provider ID for the link-layer device.
+	SetProviderIDOps(id network.Id) ([]txn.Op, error)
+}
+
+// LinkLayerAddress describes a single layer-3 network address
+// assigned to a layer-2 device.
+type LinkLayerAddress interface {
+	// DeviceName is the name of the device to which this address is assigned.
+	DeviceName() string
+
+	// Value returns the actual IP address.
+	Value() string
+
+	// Origin indicates the authority that is maintaining this address.
+	Origin() network.Origin
+
+	// SetProviderIDOps returns the operations required to set the input
+	// provider ID for the address.
+	SetProviderIDOps(id network.Id) ([]txn.Op, error)
+
+	// SetOriginOps returns the transaction operations required to change
+	// the origin for this address.
+	SetOriginOps(origin network.Origin) []txn.Op
+
+	// SetProviderNetIDsOps returns the transaction operations required to ensure
+	// that the input provider IDs are set against the address.
+	SetProviderNetIDsOps(networkID, subnetID network.Id) []txn.Op
+}
+
+// LinkLayerMachine describes a machine that can query
+// and update link-layer device data.
+type LinkLayerMachine interface {
```

Comment:
The name and docblock of this type seems confusing to me. Why not call it something like `MachineLinkLayerAccessor` or `MachineLinkLayerQuerier` etc? The method signatures indicate that whatever implements this interface only supports queries and not mutations.